pinery-service/src/main/java/ca/on/oicr/pinery/service/impl/Cache.java

```diff
@@ -0,0 +1,357 @@
+package ca.on.oicr.pinery.service.impl;
+
+import ca.on.oicr.gsi.provenance.model.LaneProvenance;
+import ca.on.oicr.gsi.provenance.model.SampleProvenance;
+import ca.on.oicr.pinery.api.AttributeName;
+import ca.on.oicr.pinery.api.Box;
+import ca.on.oicr.pinery.api.ChangeLog;
+import ca.on.oicr.pinery.api.DataProvider;
+import ca.on.oicr.pinery.api.Instrument;
+import ca.on.oicr.pinery.api.InstrumentModel;
+import ca.on.oicr.pinery.api.Lims;
+import ca.on.oicr.pinery.api.Order;
+import ca.on.oicr.pinery.api.Run;
+import ca.on.oicr.pinery.api.Sample;
+import ca.on.oicr.pinery.api.SampleProject;
+import ca.on.oicr.pinery.api.Type;
+import ca.on.oicr.pinery.api.User;
+import io.prometheus.client.Gauge;
+import io.prometheus.client.Histogram;
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.util.Date;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+
+@Service
+public class Cache implements DataProvider {
+
+  private static final Logger log = LoggerFactory.getLogger(Cache.class);
+
+  private static final Histogram cacheUpdateTime =
+      Histogram.build()
+          .name("pinery_cache_update_time")
+          .help("Time to update the cache (in seconds)")
+          .buckets(60, 180, 300, 600, 900, 1200, 1800, 3600)
+          .register();
+  private static final Gauge cacheUpdateFailures =
```

Comment:
Can you also include a gauge with the timestamp of when the cache was updated? There's a `setToCurrentTime()` method.

Reply:
added